Use Universal Developer Image as dev container

vitaliy-guliy opened the pull request:
### What does this PR do?
- Inherits che-code-dev image from Universal Developer Image
- Adds few commands to build che-code image with podman

### What issues does this PR fix?
<!-- Please include any related issue from the Eclipse Che repository (or from another issue tracker). -->
https://github.com/eclipse/che/issues/22292

### How to test this PR?
Do following:
- use the dogfooding instance to create a workspace from the current branch https://github.com/che-incubator/che-code/tree/switch-to-udi
- use devfile commands to compile and run VS Code with `yarn`
- use devfile commands to compile and then build `che-code` image with `podman`

Should have like following:

![Screenshot from 2023-06-26 17-50-43](https://github.com/che-incubator/che-code/assets/1655894/3ecc8457-7958-4ad1-b11d-028c80558320)

![Screenshot from 2023-06-26 17-49-35](https://github.com/che-incubator/che-code/assets/1655894/2c855446-9749-46a4-8108-7579d678d253)

---

### Note that:

- all the 30 commits will be squashed when merging. I haven't squashed them here because of 1) all the commits to the branch were performed using online che-code, as for me the squashing it not clear UX flow; 2) do not want to add an extra commit as it  triggers the github PR checks

- at the moment, `quay.io/che-incubator-pull-requests/che-code-dev:pr-246-dev-amd64` is used as a dev image. I plan to update it after merging of this pull request and when the corresponding che-dev image will be built by the appropriate github action.

vitaliy-guliy committed:
fix: use universal developer image as a dev container

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

vitaliy-guliy committed:
chore: temporary disabling the smoke test to build che-dev image

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

vitaliy-guliy commented:
There is one thing that needs to be fixed - is optimize the dev image as its size a bit huge
```
$ docker images
REPOSITORY                                  TAG                  IMAGE ID       CREATED         SIZE
quay.io/vgulyy/che-code-dev                 next-udi             7610232c9b24   25 hours ago    9.63GB
```

vitaliy-guliy committed:
chore: add action to build che-code-dev image

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

vitaliy-guliy committed:
chore: add action to build che-code-dev image

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

vitaliy-guliy committed:
chore: add action to build che-code-dev image

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

vitaliy-guliy committed:
chore: add action to build che-code-dev image

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

vitaliy-guliy committed:
chore: add action to build che-code-dev image

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

vitaliy-guliy committed:
chore: add action to build che-code-dev image

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
Merge branch 'main' into switch-to-udi

vitaliy-guliy committed:
chore: add action to build che-code-dev image

vitaliy-guliy committed:
temporary disable tests when build che-code image

RomanNikitenko commented on build/dockerfiles/dev.Dockerfile:
```diff
@@ -6,81 +6,49 @@
 
-# https://quay.io/eclipse/che-machine-exec#^7\.
-FROM quay.io/eclipse/che-machine-exec:7.56.0 as machine-exec
+FROM quay.io/devfile/universal-developer-image:latest
```
`universal-developer-image` has a lot of tools, but as result it has a huge image size
is it really required to have **all** those tools from the `universal-developer-image` for the `che-code` development?
can we just add required tools to the dev image?

just for my understanding, as you are working on it and you know the problem better then me...

vitaliy-guliy replied:
I had some issues with configuring the podman in current `che-dev` image.
So, It was much easier to take the universal developer image and add missing libraries, required for building vscode.

vitaliy-guliy committed:
chore: rework action to build che-code image

vitaliy-guliy committed:
chore: build che-code image pull request check

vitaliy-guliy committed:
chore: code cleanup

vitaliy-guliy committed:
chore: rename actions

vitaliy-guliy committed:
chore: revert back running of tests

azatsarynnyy reviewed (commented):
@vitaliy-guliy I believe the GitHub Workflow-related changes deserve a separate PR. As they are not directly related to the current PR. However, the current PR depends on these changes. We can review/test the GitHub Workflow simplification separately.

vitaliy-guliy commented:
> @vitaliy-guliy I believe the GitHub Workflow-related changes deserve a separate PR. As they are not directly related to the current PR. However, the current PR depends on these changes. We can review/test the GitHub Workflow simplification separately.

This pull request
- introduces the changes into che-code-dev image
- adapts pull request chesks to those changes.

Here I removed two existing jobs with messy logic (we discussed that) and added two separate independent jobs to build che-code and che-code-dev images.

azatsarynnyy commented on .github/workflows/pr-check-build-che-code-image.yaml:
```diff
@@ -0,0 +1,88 @@
+          docker buildx build \
+            --platform linux/amd64 \
+            --progress=plain \
+            -f build/dockerfiles/linux-libc.Dockerfile \
```
With the help of [matrix strategy](https://github.com/che-incubator/che-code/blob/4c94034316f9ea161fff00912fabbdd9ea653716/.github/workflows/pr-check.yml#L22), you can combine `Compile linux-libc` and `Compile linux-musl` steps into just a single step.

vitaliy-guliy replied:
I reworked it intentionaly to prevent storing somewhere (uploading artifacts) the result of two launches of the build job.
Now we do not store artifacts. Now we build libc and musl, then build che-code image, upload it to quay.io and forget.

vitaliy-guliy committed:
Merge branch 'main' into switch-to-udi

vitaliy-guliy commented:
Superseded by https://github.com/che-incubator/che-code/pull/253
